packages/twenty-server/src/engine/workspace-manager/workspace-trash-cleanup/services/workspace-trash-cleanup.service.ts

```diff
@@ -0,0 +1,176 @@
+        workspaceId,
+        schemaName,
+        tableNames,
+        trashRetentionDays,
+      );
+
+      this.logger.log(
+        `Deleted ${deletedCount} record(s) from workspace ${workspaceId}`,
+      );
+
+      return { success: true, deletedCount };
+    } catch (error) {
+      const errorMessage = error?.message || String(error);
+
+      this.logger.error(
+        `Failed to cleanup workspace ${workspaceId}: ${errorMessage}`,
+      );
+
+      return { success: false, deletedCount: 0, error: errorMessage };
+    }
+  }
+
+  private async discoverTablesWithSoftDelete(
+    schemaName: string,
+  ): Promise<string[]> {
+    const rawResults = await this.objectMetadataRepository
+      .createQueryBuilder('object')
+      .innerJoin('object.dataSource', 'dataSource')
+      .select('object.nameSingular', 'nameSingular')
+      .addSelect('object.isCustom', 'isCustom')
+      .where('dataSource.schema = :schemaName', { schemaName })
+      .andWhere('object.isActive = :isActive', { isActive: true })
+      .andWhere((qb) => {
+        const subQuery = qb
+          .subQuery()
+          .select('1')
+          .from(FieldMetadataEntity, 'field')
+          .where('field.objectMetadataId = object.id')
+          .andWhere('field.name = :fieldName', { fieldName: 'deletedAt' })
+          .andWhere('field.isActive = :isActive', { isActive: true })
+          .getQuery();
+
+        return `EXISTS ${subQuery}`;
+      })
+      .getRawMany();
+
+    return rawResults.map((row) =>
+      computeObjectTargetTable({
+        nameSingular: row.nameSingular,
+        isCustom: row.isCustom,
+      }),
+    );
+  }
+
+  private async deleteSoftDeletedRecords(
+    workspaceId: string,
+    schemaName: string,
+    tableNames: string[],
+    trashRetentionDays: number,
+  ): Promise<number> {
+    let totalDeleted = 0;
+    const cutoffDate = this.calculateCutoffDate(trashRetentionDays);
+
+    for (const tableName of tableNames) {
+      if (totalDeleted >= this.maxRecordsPerWorkspace) {
+        this.logger.log(
+          `Reached deletion limit (${this.maxRecordsPerWorkspace}) for workspace ${workspaceId}`,
+        );
+        break;
+      }
+
+      const remainingQuota = this.maxRecordsPerWorkspace - totalDeleted;
+      const deleted = await this.deleteFromTable(
+        schemaName,
+        tableName,
+        cutoffDate,
+        remainingQuota,
+      );
+
+      totalDeleted += deleted;
+    }
+
+    return totalDeleted;
+  }
+
+  private calculateCutoffDate(trashRetentionDays: number): Date {
+    const cutoffDate = new Date();
+
+    cutoffDate.setUTCHours(0, 0, 0, 0);
+    cutoffDate.setDate(cutoffDate.getDate() - trashRetentionDays + 1);
+
+    return cutoffDate;
+  }
+
+  private async deleteFromTable(
+    schemaName: string,
+    tableName: string,
+    cutoffDate: Date,
+    limit: number,
+  ): Promise<number> {
+    const result = await this.dataSource.query(
```

Comment:
Typically an example where you're going much too low in abstraction. You're loosing a lot of value by doing this.

For example in `WorkspaceDeleteQueryBuilder` which is called  when you call `delete()`, you'll see there's an `eventEmitterService`. By going "raw sql" you lose this so people won't get destroy webhooks / there will be no destroy events (maybe there's some cascade deleting happening or other parts of the app that want to be aware you destroyed records)